### PR TITLE
Compile under arch linux

### PR DESCRIPTION
**Pull request by tonylambiris**
This pull request allows for a successful compilation under Arch Linux.

### CODE REVIEW

**Comment by osqueryer** [comment]
Can one of the osquery admins reply with "ok to test" to kick off a build...

File: osquery/CMakeLists.txt [hunk 1]
```diff
@@ -98,6 +98,9 @@ ADD_OSQUERY_LINK_CORE("glog")
 ADD_OSQUERY_LINK_CORE("boost_system")
 ADD_OSQUERY_LINK_CORE("boost_filesystem")
 ADD_OSQUERY_LINK_CORE("boost_regex")
+ADD_OSQUERY_LINK_CORE("cryptsetup")
```
Comment (theopolis) [inline osquery/CMakeLists.txt]: I don't think the core "libosquery.a" needs to link cryptsetup. Why is this included? The same applies to bz2 and lz4, but I could see either RocksDB or another core library wanting these. However these have not needed core linkage thus far.

**Comment by theopolis** [comment]
ok to test

**Comment by osqueryer** [comment]
Build origin/pr/2238/merge (Job: 2859) failed :-1:

File: osquery/CMakeLists.txt (hunk 1 above)
Comment (DaanDeMeyer) [inline osquery/CMakeLists.txt]: I've just tested this. Building the core (make sdk) does indeed not require these libraries. A normal build (make) does require them.

Comment (tonylambiris) [inline osquery/CMakeLists.txt]: @theopolis I get undefined references to crypt_() functions. I can provide extended output if needed.

**Comment by theopolis** [comment]
If you give me a few more days I may have a solution to help with the Arch build. It requires a redesign of the build provisioning which is about 80% finished.

**Comment by tonylambiris** [comment]
@theopolis awesome!

In case this helps, here are the errors I get when using my previous pull request but not this one:
```
[ 77%] Linking CXX executable osqueryi
/usr/local/lib/libaws-cpp-sdk-core.a(FileSystemUtils.cpp.o): In function `Aws::Utils::FileSystemUtils::CreateTempFilePath[abi:cxx11]()':
FileSystemUtils.cpp:(.text+0x1e1b): warning: the use of `tmpnam' is dangerous, better use `mkstemp'
/usr/bin/../lib64/gcc/x86_64-pc-linux-gnu/6.1.1/../../../../lib64/librocksdb.a(format.o): In function `rocksdb::UncompressBlockContents(char const*, unsigned long, rocksdb::BlockContents*, unsigned int)':
(.text+0x1435): undefined reference to `LZ4_decompress_safe'
/usr/bin/../lib64/gcc/x86_64-pc-linux-gnu/6.1.1/../../../../lib64/librocksdb.a(format.o): In function `rocksdb::UncompressBlockContents(char const*, unsigned long, rocksdb::BlockContents*, unsigned int)':
(.text+0x14a9): undefined reference to `LZ4_decompress_safe'
libosquery_additional.a(disk_encryption.cpp.o): In function `osquery::tables::genFDEStatusForBlockDevice(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::vector<std::map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > >, std::allocator<std::map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > > > >&)':
/home/tlambiris/osquery/osquery/tables/system/linux/disk_encryption.cpp:(.text._ZN7osquery6tables26genFDEStatusForBlockDeviceERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEES8_RSt6vectorISt3mapIS6_S6_St4lessIS6_ESaISt4pairIS7_S6_EEESaISG_EE+0x186): undefined reference to `crypt_status'
/home/tlambiris/osquery/osquery/tables/system/linux/disk_encryption.cpp:(.text._ZN7osquery6tables26genFDEStatusForBlockDeviceERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEES8_RSt6vectorISt3mapIS6_S6_St4lessIS6_ESaISt4pairIS7_S6_EEESaISG_EE+0x20b): undefined reference to `crypt_init_by_name_and_header'
/home/tlambiris/osquery/osquery/tables/system/linux/disk_encryption.cpp:(.text._ZN7osquery6tables26genFDEStatusForBlockDeviceERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEES8_RSt6vectorISt3mapIS6_S6_St4lessIS6_ESaISt4pairIS7_S6_EEESaISG_EE+0x21d): undefined reference to `crypt_get_type'
/home/tlambiris/osquery/osquery/tables/system/linux/disk_encryption.cpp:(.text._ZN7osquery6tables26genFDEStatusForBlockDeviceERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEES8_RSt6vectorISt3mapIS6_S6_St4lessIS6_ESaISt4pairIS7_S6_EEESaISG_EE+0x268): undefined reference to `crypt_get_active_device'
/home/tlambiris/osquery/osquery/tables/system/linux/disk_encryption.cpp:(.text._ZN7osquery6tables26genFDEStatusForBlockDeviceERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEES8_RSt6vectorISt3mapIS6_S6_St4lessIS6_ESaISt4pairIS7_S6_EEESaISG_EE+0x27a): undefined reference to `crypt_get_cipher'
/home/tlambiris/osquery/osquery/tables/system/linux/disk_encryption.cpp:(.text._ZN7osquery6tables26genFDEStatusForBlockDeviceERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEES8_RSt6vectorISt3mapIS6_S6_St4lessIS6_ESaISt4pairIS7_S6_EEESaISG_EE+0x2af): undefined reference to `crypt_get_cipher_mode'
/home/tlambiris/osquery/osquery/tables/system/linux/disk_encryption.cpp:(.text._ZN7osquery6tables26genFDEStatusForBlockDeviceERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEES8_RSt6vectorISt3mapIS6_S6_St4lessIS6_ESaISt4pairIS7_S6_EEESaISG_EE+0x4b7): undefined reference to `crypt_free'
/home/tlambiris/osquery/osquery/tables/system/linux/disk_encryption.cpp:(.text._ZN7osquery6tables26genFDEStatusForBlockDeviceERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEES8_RSt6vectorISt3mapIS6_S6_St4lessIS6_ESaISt4pairIS7_S6_EEESaISG_EE+0x54e): undefined reference to `crypt_free'
/usr/bin/../lib64/gcc/x86_64-pc-linux-gnu/6.1.1/../../../../lib64/librocksdb.a(block_based_table_builder.o): In function `rocksdb::BlockBasedTableBuilder::WriteBlock(rocksdb::Slice const&, rocksdb::BlockHandle*)':
(.text+0xb5c): undefined reference to `LZ4_compressBound'
/usr/bin/../lib64/gcc/x86_64-pc-linux-gnu/6.1.1/../../../../lib64/librocksdb.a(block_based_table_builder.o): In function `rocksdb::BlockBasedTableBuilder::WriteBlock(rocksdb::Slice const&, rocksdb::BlockHandle*)':
(.text+0xba9): undefined reference to `LZ4_compressHC2_limitedOutput'
/usr/bin/../lib64/gcc/x86_64-pc-linux-gnu/6.1.1/../../../../lib64/librocksdb.a(block_based_table_builder.o): In function `rocksdb::BlockBasedTableBuilder::WriteBlock(rocksdb::Slice const&, rocksdb::BlockHandle*)':
(.text+0xc18): undefined reference to `LZ4_compressBound'
/usr/bin/../lib64/gcc/x86_64-pc-linux-gnu/6.1.1/../../../../lib64/librocksdb.a(block_based_table_builder.o): In function `rocksdb::BlockBasedTableBuilder::WriteBlock(rocksdb::Slice const&, rocksdb::BlockHandle*)':
(.text+0xc5e): undefined reference to `LZ4_compress_limitedOutput'
clang-3.8: error: linker command failed with exit code 1 (use -v to see invocation)
make[3]: *** [osquery/CMakeFiles/shell.dir/build.make:151: osquery/osqueryi] Error 1
make[2]: *** [CMakeFiles/Makefile2:604: osquery/CMakeFiles/shell.dir/all] Error 2
make[1]: *** [Makefile:139: all] Error 2
make: *** [Makefile:32: all] Error 2
```

**Comment by tonylambiris** [comment]
The AUR package I've been working on is now live!
https://aur.archlinux.org/packages/osquery-git

Also I created an AUR package for aws-sdk-cpp as well as to not cruft up /usr/local:
https://aur.archlinux.org/packages/aws-sdk-cpp-git/

@DaanDeMeyer I tried rebuilding rocksdb with the exported variable from the AUR but it didn't seem to include the statically compiled library. Were you able to get this to work?

File: osquery/CMakeLists.txt (hunk 1 above)
Comment (theopolis) [inline osquery/CMakeLists.txt]: Yeah, if you don't mind. I'd like to see the linking errors. There shouldn't be anything in osquery "core" (essentially the non-table, non-plugin) code that requires `libcryptsetup` symbols.

**Comment by DaanDeMeyer** [comment]
@tonylambiris I'm not exactly sure what you mean. Are you referring to BUILD_LINK_SHARED? Right now you're listing rocksdb and rocksdb-static as dependencies. Osquery will only use the static library (by default) but can switch to the shared library with BUILD_LINK_SHARED if both versions are installed. I suggest using the shared library since all the other libraries are shared as well. One thing I'm not sure about is if osquery automatically detects the shared rocksdb library. You might still have to explicitly build with BUILD_LINK_SHARED=True.

**Comment by theopolis** [comment]
That environment variable flips the precendence for linking. But if you only have the shared libraries available, those will be used to link :)

**Comment by tonylambiris** [comment]
@DaanDeMeyer ah sorry I didn't realize the BUILD_LINK_SHARED variable was for osquery, I was confusing that with building rocksdb. FWIW I have both rocksdb and rocksdb-static both installed on my Arch host and there aren't any conflicts with having both installed. 👍

**Comment by osqueryer** [comment]
Can one of the osquery admins reply with "ok to test" to kick off a build...

**Comment by theopolis** [comment]
Alright! The new build system has been merged and I'm ready to deal with the regressions and fallout. ;)

To test if these changes are still needed:
```
make distclean
make deps
make
```

**Comment by theopolis** [comment]
I'll close this out for now, if the new build system doesn't work (or if we need some edits) let's reopen or track in a new PR. 💃
